code/Form/UserFormsRequiredFields.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+namespace SilverStripe\UserForms\Form;
+
+use InvalidArgumentException;
+use SilverStripe\Dev\Debug;
+use SilverStripe\Forms\FileField;
+use SilverStripe\Forms\FormField;
+use SilverStripe\Forms\RequiredFields;
+use SilverStripe\ORM\ArrayLib;
+use SilverStripe\UserForms\Model\EditableFormField;
+
+/**
+ * An extension of RequiredFields which handles conditionally required fields.
+ *
+ * A conditionally required is a field that is required, but can be hidden by display rules.
+ * When it is visible, (according to the submitted form data) it will be validated as required.
+ * When it is hidden, it will skip required validation.
+ *
+ * Required fields will be validated as usual.
+ * Conditionally required fields will be validated IF the display rules are satisfied in the submitted dataset.
+ */
+class UserFormsRequiredFields extends RequiredFields
+{
+    /**
+     * Allows validation of fields via specification of a php function for
+     * validation which is executed after the form is submitted.
+     *
+     * @param array $data
+     *
+     * @return bool
+     */
+    public function php($data)
+    {
+        $valid = true;
+        $fields = $this->form->Fields();
+
+        foreach ($fields as $field) {
+            $valid = ($field->validate($this) && $valid);
```

Comment:
could the order on this be ($valid && $field->validate($this)) ? or would do this mess up the validation messages?

Reply:
This will mess up the validation. `FormField::validate()` is meant to attach an error message to the validator if it fails. So if have 2 fields with invalid data and you start with checking the state of `valid`, the second field won't get flag has invalid which will be confusing for the user.

Here's an exmaple of a typical FormField validate method.
https://github.com/silverstripe/silverstripe-framework/blob/547ac6f2ce3870611ed60a2b8ce18ab7a8808005/src/Forms/EmailField.php#L30-L50